### DISCUSSION TITLE
Batch Sensor working with PTracer

### DISCUSSION THREAD

**Discussion by Ghiraa** [mitsuba-renderer/mitsuba3#1090] [Q&A]
Is there a method to utilize a batch sensor with the "ptracer" integrator in Mitsuba without creating a custom sensor? If this isn't possible, could you recommend any examples of custom sensors for guidance? Similarly, I'm interested in developing a custom integrator similar to "ptracer" that also considers participating media, akin to "volpath". As I'm new to Mitsuba, any advice or direction on where to find relevant resources or examples would be greatly appreciated. Thanks!

**Comment by njroussel** [comment]
Hi @Ghiraa 

I believe `ptracer` should work with `batch` sensor. Are you getting an error message?

As for a custom integrator, in addition to the `ptracer` I would recommend having a look at `prbvolpath.sample()` and `prb_projective.sample_importance()`. The former should help you with the Python interface of the medium sampling and the latter with the light tracing style of rendering to connect your path to a sensor. 
Other than that, [PBRT](https://pbr-book.org/4ed/contents) should guide you through the actual volume rendering and light tracing theory if you're unfamiliar with it.

**Reply by Ghiraa** [reply]
If set the integrator to ptracer, I just get a black output with no error message. I have been experimenting with some sensors around the cube, I have attached some screenshots.

This is what I get if I render from each sensor individually and when I render using the batch sensor:
![image](https://github.com/mitsuba-renderer/mitsuba3/assets/45585468/3dc8b14b-96e8-49bc-a491-9bf6f7188a10)

![image](https://github.com/mitsuba-renderer/mitsuba3/assets/45585468/b59824ef-f26f-4c8f-b958-a763b0bb2649)

And this is what I get when I switch to ptracer:
![image](https://github.com/mitsuba-renderer/mitsuba3/assets/45585468/f2188095-0f84-4aa7-ba05-0a61241ff04a)

![image](https://github.com/mitsuba-renderer/mitsuba3/assets/45585468/59bb4164-713b-4bb9-8285-02c84f37162d)

So, I am doing something wrong with my code, there are hidden errors or batch sensor does not support "ptracer", which makes me think I might need a custom batch sensor? Thank you!

**Reply by njroussel** [reply]
Thank you for the details. This sounds like a bug,  we'll have a look.

**Reply by njroussel** [reply]
Hi @Ghiraa 

I'm unable to reproduce what you have described here. Please open an issue if this persists.

Here is what I did: 

```python
import mitsuba as mi
import drjit as dr

mi.set_variant('cuda_ad_rgb')

scene_description = mi.cornell_box()
scene_description['integrator']['type'] = 'ptracer'
original_sensor = dict(scene_description['sensor'])

N = 5
# Programatically build N sensor dictionnaries (translated vertically)
sensors = []
for i in range(N):
    translate_y = 0.2 - i * 0.1
    tf = mi.ScalarTransform4f().translate([0, translate_y, 0]) @ original_sensor['to_world']
    new_sensor_i = dict(original_sensor)
    new_sensor_i['to_world'] = tf
    sensors.append(new_sensor_i)

# Update film width
original_film = dict(scene_description['sensor']['film'])
original_film['width'] = 256 * N

# Programatically build `batch` sensor
sensor_dict = {
    'type': 'batch',
    'film': original_film
}
for i in range(len(sensors)):
    sensor_dict[f'sensor_{i}'] = sensors[i]

sensor = mi.load_dict(sensor_dict)
mi.Bitmap(mi.render(mi.load_dict(scene_description), sensor=sensor)).write('tmp.exr')
```

If you file an issue, please build a minimal reproducer like I tried above (using only `mi.cornell_box()` for the scene resources). Thank you in advance :)

**Reply by Ghiraa** [reply]
I can't run the code above, my kernel keeps dying for some reason.

**Reply by merlinND** [reply]
@Ghiraa If you are encountering a crash, please run the code as a  Python script in the terminal and report which error is printed before the crash.

**Reply by Ghiraa** [reply]
Hi, sorry for coming back so late to this post. I ran it as a python script and the issue is that I could not use the "'cuda_ad_rgb'", but after changing it it worked perfectly fine. I still wonder what's issue with my code though, here is how I create them:
```py
def load_sensor(r, phi, theta):
    # Apply two rotations to convert from spherical coordinates to world 3D coordinates.
    origin = T.rotate([0, 0, 1], phi).rotate([0, 1, 0], theta) @ mi.ScalarPoint3f([0, 0, r])

    return {
        'type': 'perspective',
        'fov': 39.3077,
        'to_world': T.look_at(
            origin=origin,
            target=[0, 0, 0],
            up=[0, 0, 1]
        ),
        'sampler': {
            'type': 'independent',
            'sample_count': 16
        },
        'film': {
            'type': 'hdrfilm',
            'width': 256,
            'height': 256,
            'rfilter': {
                'type': 'tent',
            },
            'pixel_format': 'rgb',
        },
    }

sensor_description = {
    'type': 'batch',
    'sampler': {
            'type': 'independent',
            'sample_count': 16
    },
    'film': {
        'type': 'hdrfilm',
        'width': 6 * 256,
        'height': 254,
        'rfilter': {
            'type': 'tent',
        },
        'pixel_format': 'rgb',
    },
}

sensor_count = 6

radius = 6
phis = [60.0 * i for i in range(sensor_count)]
theta = 60.0
for i, phi in enumerate(phis):
    sensor_description[f"persp_{i}"] = load_sensor(radius, phi, theta)

sensorr = mi.load_dict(sensor_description)
image = mi.render(scene, spp=16, sensor=sensorr)

sensors = [mi.load_dict(load_sensor(radius, phi, theta)) for phi in phis]
```

**Reply by merlinND** [reply]
Glad to hear that you got it working. If it works, then what do you mean by "I still wonder what's issue with my code though"?
What is not going as expected?

**Reply by njroussel** [reply]
@Ghiraa There's a typo in your code:

``` 
'film': {
        'type': 'hdrfilm',
        'width': 6 * 256,
        'height': 254, # <--- HERE
        'rfilter': {
            'type': 'tent',
        },
        'pixel_format': 'rgb',
    },
 ```
 
 The `height` should be 256. I checked the source code, we never actually check that the film's height is consistent with the "sub"-sensors'.
